[bug] compiler = 'msvc' does not work with Ninja generator

It is currently not use Ninja as generator in combination with the new `msvc` compiler. This combination causes CMake to abort with an error (see below).

The error appears because the generated toolchain file sets `CMAKE_GENERATOR_TOOLSET` (see [cmake docs](https://cmake.org/cmake/help/latest/variable/CMAKE_GENERATOR_TOOLSET.html) which is only supported for `Visual Studio *` generators. Extending this block to set `toolset=None` the CMake stops complaining:
https://github.com/conan-io/conan/blob/9ebc2e4282a35e4d29a7b697d71b85db45c31836/conan/tools/cmake/toolchain.py#L579

However, simply removing the block in the toolchain file does not solve the problem entirely because the generated `conanvcvars.bat` does not configure a specific version of the compiler (as indicated by configuring Visual Studio 16.0 without specifying the v141 toolset).

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10 20H2
  * Compiler+version: msvc 19.2
  * Conan version: 1.37.2
  * Python version: Bundled with Conan (I'm using the chocolatey package)

### Steps to reproduce (Include if Applicable)
Here are three files to reproduce this locally:

**conanfile.py**
```python
from conans import ConanFile, tools
from conan.tools.cmake import CMake, CMakeToolchain, CMakeDeps


class MWEParser(ConanFile):
    name = "MWE"
    version = "0.0.1"
    settings = "os", "build_type", "compiler", "arch"
    exports_sources = "CMakeLists.txt"

    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()
        deps = CMakeDeps(self)
        deps.generate()

    def configure_cmake(self):
        cmake = CMake(self)
        cmake.configure()
        return cmake

    def build(self):
        cmake = self.configure_cmake()
        cmake.build()

    def package(self):
        cmake = self.configure_cmake()
        cmake.install()
```

**profile**
```ini
[settings]
os=Windows
arch=x86_64
compiler=msvc
compiler.version=19.1
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.cppstd=17
build_type=Release
[conf]
tools.cmake.cmaketoolchain:generator=Ninja
tools.microsoft.msbuild:vs_version = 16
```

**CMakeLists.txt**
```cmake
project(MWE)
```


### Logs (Executed commands with output) (Include/Attach if Applicable)

`conan create . -pr .\profile`
```log
Installing (downloading, building) binaries...
MWE/0.0.1: Configuring sources in D:\dev\.conan\data\MWE\0.0.1\_\_\source
MWE/0.0.1: Copying sources to build folder
MWE/0.0.1: Building your package in D:\dev\.conan\data\MWE\0.0.1\_\_\build\b7ed69139ab470a06c78efe6528bfbaa5ca4c3bb
MWE/0.0.1: Generator txt created conanbuildinfo.txt
MWE/0.0.1: Calling generate()
MWE/0.0.1: Calling build()
MWE/0.0.1: CMake command: cmake -G "Ninja" -DCMAKE_TOOLCHAIN_FILE="D:/dev/.conan/data/MWE/0.0.1/_/_/build/b7ed69139ab470a06c78efe6528bfbaa5ca4c3bb/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="D:/dev/.conan/data/MWE/0.0.1/_/_/package/b7ed69139ab470a06c78efe6528bfbaa5ca4c3bb" "D:\dev\.conan\data\MWE\0.0.1\_\_\build\b7ed69139ab470a06c78efe6528bfbaa5ca4c3bb"
**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.6.4
** Copyright (c) 2020 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'
Using Conan toolchain through D:/dev/.conan/data/MWE/0.0.1/_/_/build/b7ed69139ab470a06c78efe6528bfbaa5ca4c3bb/conan_toolchain.cmake.
-- Conan C++ Standard 17 with extensions OFF}
  Generator

    Ninja

  does not support toolset specification, but toolset

    v142

  was specified.


CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
```

Generated `conanvcvars.bat`:
```bat
@echo off
set "VSCMD_START_DIR=%CD%" && call "C:\Program Files (x86)\Microsoft Visual Studio 16.0\VC/Auxiliary/Build/vcvarsall.bat"  amd64
```

--- a/conan/tools/cmake/toolchain.py
+++ b/conan/tools/cmake/toolchain.py
@@ -483,6 +483,8 @@ class GenericSystemBlock(Block):
         """)
 
     def _get_toolset(self, generator):
+        if generator is None or ("Visual" not in generator and "Xcode" not in generator):
+            return None
         settings = self._conanfile.settings
         compiler = settings.get_safe("compiler")
         compiler_base = settings.get_safe("compiler.base")

--- a/conan/tools/microsoft/toolchain.py
+++ b/conan/tools/microsoft/toolchain.py
@@ -26,8 +26,23 @@ def write_conanvcvars(conanfile):
     elif compiler == "Visual Studio" or compiler == "msvc":
         vs_version = vs_ide_version(conanfile)
         vcvarsarch = vcvars_arch(conanfile)
+        vcvars_ver = None
+        if compiler == "Visual Studio":
+            toolset = conanfile.settings.get_safe("compiler.toolset")
+            if toolset is not None:
+                vcvars_ver = {"v140": "14.0",
+                              "v141": "14.1",
+                              "v142": "14.2"}.get(toolset)
+        else:
+            # Code similar to CMakeToolchain toolset one
+            compiler_version = str(conanfile.settings.compiler.version)
+            version_components = compiler_version.split(".")
+            assert len(version_components) >= 2  # there is a 19.XX
+            minor = version_components[1]
+            # The equivalent of compiler 19.26 is toolset 14.26
+            vcvars_ver = "14.{}".format(minor)
         cvars = vcvars_command(vs_version, architecture=vcvarsarch, platform_type=None,
-                               winsdk_version=None, vcvars_ver=None)
+                               winsdk_version=None, vcvars_ver=vcvars_ver)
     if cvars:
         content = textwrap.dedent("""\
             @echo off

--- a/conans/test/functional/toolchains/cmake/test_ninja.py
+++ b/conans/test/functional/toolchains/cmake/test_ninja.py
@@ -117,6 +117,41 @@ def test_locally_build_msvc(build_type, shared, client):
 
 
 @pytest.mark.skipif(platform.system() != "Windows", reason="Only windows")
+@pytest.mark.tool_compiler
+@pytest.mark.tool_ninja
+def test_locally_build_msvc_toolset(client):
+    msvc_version = "15"
+    profile = textwrap.dedent("""
+        [settings]
+        os=Windows
+        compiler=msvc
+        compiler.version=19.0
+        compiler.runtime=dynamic
+        compiler.cppstd=14
+        build_type=Release
+        arch=x86_64
+        [conf]
+        tools.cmake.cmaketoolchain:generator=Ninja
+        tools.microsoft.msbuild:vs_version = 15
+        """)
+    client.save({"profile": profile})
+    client.run("install . -pr=profile")
+
+    client.run_command('conanvcvars.bat && cmake . -G "Ninja" '
+                       '-DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake '
+                       '-DCMAKE_BUILD_TYPE=Release')
+
+    client.run_command("conanvcvars.bat && ninja")
+
+    client.run_command("myapp.exe")
+
+    # Checking that compiler is indeed version 19.0, not 19.1-default of VS15
+    check_exe_run(client.out, ["main", "hello"], "msvc", "19.0", "Release", "x86_64", cppstd="14")
+    check_vs_runtime("myapp.exe", client, msvc_version, "Release", architecture="amd64")
+    check_vs_runtime("mylibrary.lib", client, msvc_version, "Release", architecture="amd64")
+
+
+@pytest.mark.skipif(platform.system() != "Windows", reason="Only windows")
 @pytest.mark.parametrize("build_type,shared", [("Release", False), ("Debug", True)])
 @pytest.mark.tool_mingw64
 @pytest.mark.tool_compiler
@@ -186,6 +221,9 @@ def test_ninja_conf():
     client.run("install . -pr=profile")
     conanbuild = client.load("conanbuild.json")
     assert '"cmake_generator": "Ninja"' in conanbuild
-    if platform.system() == "Windows":
-        vcvars = client.load("conanvcvars.bat")
-        assert "2017" in vcvars
+    vcvars = client.load("conanvcvars.bat")
+    assert "2017" in vcvars
+
+    # toolchain cannot define the CMAKE_GENERATOR_TOOLSET for Ninja
+    cmake = client.load("conan_toolchain.cmake")
+    assert "CMAKE_GENERATOR_TOOLSET" not in cmake

